#5417: Pinned location sharing

mnaturel opened the pull request:
<!-- Please read [CONTRIBUTING.md](https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md) before submitting your pull request -->
 
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

<!-- Describe shortly what has been changed -->
Implementation of the pinned location sharing. User can now select a location on the map different from its current location. User can also zoom to its current location using a dedicated button placed above the compass icon.

## Motivation and context

<!-- Provide link to the corresponding issue if applicable or explain the context -->
Closes #5417 

## Screenshots / GIFs

<!-- Only if UI have been changed -->
### User location sharing
<img src="https://user-images.githubusercontent.com/46314705/157465169-4e0cfb52-c7c4-43cd-8495-8efcaaeff8b5.gif" width=300 />

### Pinned location sharing
<img src="https://user-images.githubusercontent.com/46314705/157464541-2a6b42cb-90f2-4007-aea8-50cee8f16392.gif" width=300 />

## Tests

<!-- Explain how you tested your development -->

User location case
- Ensure location sharing option is enabled in "Settings" -> "Preferences"
- Go to a room or a direct message
- Click on the "add" icon
- Select the location icon
- Check you see the map with your current location
- Check you see the confirmation view with your avatar and color
- Click on the confirmation view
- Check the location is correctly shared in the timeline of the room/message

Pinned location case
- Ensure location sharing option is enabled in "Settings" -> "Preferences"
- Go to a room or a direct message
- Click on the "add" icon
- Select the location icon
- Check you see the map with your current location
- Move the map on another location
- Check the confirmation view has changed and corresponds to pin location option
- Check the pin on the map is now a generic pin instead of the avatar
- Click on the confirmation view
- Check the location is correctly shared in the timeline of the room/message

## Tested devices

- [x] Physical
- [x] Emulator
- OS version(s): Android 11

## Checklist

<!-- Depending on the Pull Request content, it can be acceptable if some of the following checkboxes stay unchecked. -->

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

gaelledel reviewed (approved):
Awesome work :D

onurays commented on vector/src/main/java/im/vector/app/features/location/LocationSharingAction.kt:
```diff
@@ -19,5 +19,8 @@ package im.vector.app.features.location
+    object CurrentUserLocationSharingAction : LocationSharingAction()
+    data class PinnedLocationSharingAction(val locationData: LocationData?) : LocationSharingAction()
+    data class LocationTargetChangeAction(val locationData: LocationData) : LocationSharingAction()
+    object ZoomToUserLocationAction : LocationSharingAction()
```
Can you remove Action suffix like `ZoomToUserLocation` to continue the naming convention of other actions?

onurays commented on vector/src/main/java/im/vector/app/features/location/LocationSharingFragment.kt:
```diff
@@ -130,21 +143,52 @@ class LocationSharingFragment @Inject constructor(
                 .show()
     }
 
+    private fun initLocateBtn() {
```
I would prefer renaming to `initLocateButton()`

onurays commented on vector/src/main/java/im/vector/app/features/location/LocationSharingViewModel.kt:
```diff
@@ -47,50 +65,106 @@ class LocationSharingViewModel @AssistedInject constructor(
-                        pinDrawable = it
-                )
+    private fun updatePin(isUserPin: Boolean? = true) {
+        if (isUserPin == true) {
```
We generally prefer `if (isUserPin.orFalse())`.

mnaturel replied:
Thanks, I was not aware of this extension.

onurays commented on vector/src/main/java/im/vector/app/features/location/LocationSharingViewState.kt:
```diff
@@ -43,7 +44,8 @@ data class LocationSharingViewState(
         pinId = DEFAULT_PIN_ID,
-        pinDrawable = pinDrawable
+        pinDrawable = null,
+        showPin = areTargetAndUserLocationEqual == false
```
`.orFalse` can be used here too.

mnaturel replied:
Actually I replaced it by `areTargetAndUserLocationEqual.orTrue().not()` to be equivalent to `areTargetAndUserLocationEqual == false`.

onurays reviewed (changes requested):
Nice work! I just have small remarks.

mnaturel commented on vector/src/main/java/im/vector/app/features/location/MapTilerMapView.kt:
```diff
@@ -42,24 +50,100 @@ class MapTilerMapView @JvmOverloads constructor(
+    val locateBtn by lazy { createLocateBtn() }
     private var mapRefs: MapRefs? = null
     private var initZoomDone = false
+    private var showLocationBtn = false
```
Yes no problem, done.

mnaturel commented:
> Nice work! Tested on my device. I just have small remarks.

Thanks! I fixed all comments and rebased the branch on develop. Can you have a look again?
